File: python/01_contracts/withdraw.py
"""Lesson 1: executable preconditions and postconditions."""


def withdraw(balance: int, amount: int) -> int:
    """Return the balance after withdrawing ``amount``.

    Raises:
        ValueError: if a precondition is violated.
        AssertionError: if this implementation violates a postcondition.

    Type hints document the intended input type. Unlike Scala's types, Python
    does not enforce these hints at runtime.
    """
    if balance < 0:
        raise ValueError("balance must be non-negative")
    if amount < 0:
        raise ValueError("amount must be non-negative")
    if amount > balance:
        raise ValueError("amount must not exceed balance")

    new_balance = balance - amount

    # These are implementation promises, not checks on the caller.
    if new_balance < 0:
        raise AssertionError("postcondition failed: result must be non-negative")
    if new_balance != balance - amount:
        raise AssertionError("postcondition failed: result must equal balance - amount")

    return new_balance

File: scala/01_contracts/src/main/scala/Withdraw.scala
package contracts

object Withdraw:
  enum WithdrawalError:
    case NegativeBalance
    case NegativeAmount
    case AmountExceedsBalance

  def withdraw(balance: Int, amount: Int): Either[WithdrawalError, Int] =
    if balance < 0 then Left(WithdrawalError.NegativeBalance)
    else if amount < 0 then Left(WithdrawalError.NegativeAmount)
    else if amount > balance then Left(WithdrawalError.AmountExceedsBalance)
    else
      val newBalance = balance - amount

      // A failed postcondition is an implementation defect, not a domain error.
      Right(
        newBalance.ensuring(_ >= 0, "result must be non-negative")
          .ensuring(_ == balance - amount, "result must equal balance - amount")
      )
